protocols/ping/src/handler.rs

```diff
@@ -21,107 +21,207 @@
 use crate::protocol;
 use futures::prelude::*;
 use libp2p_core::ProtocolsHandlerEvent;
-use libp2p_core::protocols_handler::{KeepAlive, OneShotHandler, ProtocolsHandler, ProtocolsHandlerUpgrErr};
-use std::{io, time::Duration, time::Instant};
+use libp2p_core::protocols_handler::{
+    KeepAlive,
+    ProtocolsHandler,
+    ProtocolsHandlerUpgrErr,
+};
+use std::{io, num::NonZeroU32, time::{Duration, Instant}};
 use tokio_io::{AsyncRead, AsyncWrite};
 use tokio_timer::Delay;
+use void::Void;
 
-/// Protocol handler that handles pinging the remote at a regular period and answering ping
-/// queries.
+/// The configuration for outbound pings.
+#[derive(Clone)]
+pub struct PingConfig {
+    /// The timeout of an outbound ping.
+    timeout: Duration,
+    /// The duration between the last successful outbound or inbound ping
+    /// and the next outbound ping.
+    interval: Duration,
+    /// The maximum number of failed outbound pings before the associated
+    /// connection is deemed unhealthy, indicating to the `Swarm` that it
+    /// should be closed.
+    max_timeouts: NonZeroU32
```

Comment:
What's the reason for having a value different from `1` here?
You mention that inbound pings can keep a connection alive even if outbound pings fail, but I don't really see how that could happen unless the remote is malicious/buggy/overloaded.

Reply:
> What's the reason for allowing values different from `1` here? (EDIT: by that I mean: what's the reason for having this feature at all?)

I think basing a connection keep-alive on the timeout of a single request is a bad idea and leads to unnecessary disconnects / connection turnover. There can always be the occasional unfortunate request that gets routed along a bad path. Don't you think so?

> You mention that inbound pings can keep a connection alive even if outbound pings fail, but I don't really see how that could happen unless the remote is malicious/buggy/overloaded.

That is simply because, as I mentioned, when a ping is received that currently also resets the local node's own ping timer and extends the keep-alive of the connection. I.e. every inbound ping or answer to an outbound ping is taken as indication that the connection is still "alive". Only if there is a time window where pings don't get through in either direction is the connection deemed "dead".

Reply:
> There can always be the occasional unfortunate request that gets routed along a bad path. Don't you think so?

All the libp2p transports must be reliable, so if the ping doesn't go through it will be sent again.

Reply:
> > There can always be the occasional unfortunate request that gets routed along a bad path. Don't you think so?
> 
> This is true for ICMP pings, but here all libp2p transports must be reliable, so if the ping doesn't go through it will automatically be sent again.

Hm, the semantics of "reliable" are necessarily not well-defined for the transports though and there can be other reasons in the sending or receiving application for an occasional transient error during processing of a request, which I think should not immediately cause a connection to be closed. Thus I still think this configuration option is useful and should not default to `1`. Does anyone else reading this have an opinion on the matter one way or the other?

Reply:
>  the semantics of "reliable" are necessarily not well-defined for the transports though

It's probably not well documented, but implementations of `StreamMuxer` must be reliable (either the connection is well-working or it's completely dead, there's no in between), and a transport must be reliable if we build a muxer on top of it.

> there can be other reasons in the sending or receiving application for an occasional transient error during processing of a request

I'd argue that no, there shouldn't be any reason for an occasional transient error!

Reply:
> > the semantics of "reliable" are necessarily not well-defined for the transports though
> 
> It's probably not well documented, but implementations of `StreamMuxer` must be reliable (either the connection is well-working or it's completely dead, there's no in between), and a transport must be reliable if we build a muxer on top of it.

I meant something else though, namely that the definition of "reliable" is never the same between two different transports, e.g. it has a very specific meaning for TCP in terms of maximum timeouts for retransmissions before giving up, and many other configuration options. I think the common denominator for the abstract concept of an underlying reliable transport in the context of a higher-level protocol as in `libp2p-ping` can only be that it has some form of error-recovery e.g. through retransmission and/or error-detecting or error-correcting codes and delivers messages in-order, if they are delivered at all.

> > there can be other reasons in the sending or receiving application for an occasional transient error during processing of a request
> 
> I'd argue that no, there shouldn't be any reason for an occasional transient error!

That would be great, but does not match my perception of reality, where even assuming a flawless transport and network, there are a lot of other pieces of software and hardware involved on both the sender's and receiver's side, all subject to transient failures of varying frequency. All I'm saying is that I don't think it is advisable to judge a network connection (or really anything) as broken based on a single observation of failure, given that a) they are not cheap to recreate and b) it can cause significant work for application protocols to pick up from where they left off on a new connection.

I'm not sure its worth so much discussion though, given that the current approach certainly allows the configuration that you prefer. Are you strongly opposed to allowing this to be configurable at all?

Reply:
> All I'm saying is that I don't think it is advisable to judge a network connection (or really anything) as broken based on a single observation of failure

I'm pretty sure that TCP guarantees that no packet gets lost or modified.

Additionally, the state of the stream cipher in the symmetric encryption follows the data that we send on the connection. If any byte we send is lost the connection become unrecoverable, even if we wanted to recover from it. The hmac also guarantees that if any byte is modified, an error is  triggered and the whole connection gets killed as well.

> I'm not sure its worth so much discussion though, given that the current approach certainly allows the configuration that you prefer. Are you strongly opposed to allowing this to be configurable at all?

I'm not opposed to having this configurable. I think the default should be `1` though.